keras/engine/topology.py

```diff
@@ -3063,6 +3063,29 @@ def preprocess_weights_for_loading(layer, weights,
             # split the bias into half and merge
             weights[2] = bias[:units * 4] + bias[units * 4:]
 
+    # convert the weights of CuDNNGRU so that they could be loaded into
+    if layer.__class__.__name__ == 'GRU' and len(weights) == 3:
+        # determine if we're loading a CuDNNGRU layer from the number of bias weights:
+        # CuDNNGRU has (units * 6) weights; while GRU has (units * 3)
+        # if there's no bias weight in the file, skip this conversion
+        # TODO: refactor with the LSTM conversion code above
+        units = weights[1].shape[0]
+        bias = weights[2]
+        n_gates = 3
+        if len(bias) == units * n_gates * 2:
```

Comment:
What if we're loading a `GRU` layer with `reset_after=True`? It will have 6 biases also. I think we shouldn't transpose the kernels in this case.

Reply:
It might be the cause of failed test:
```
tests/keras/engine/test_topology.py::test_preprocess_weights_for_loading[layer0] FAILED
```

Reply:
This is a really good point. There's a problem that the loading code has no clue how to distinguish weights from `CuDNNGRU` and `GRU(reset_after=True)` since they have the same shape and there are not metadata in the HDF file.

Reply:
I can see some ways how to make those cases distinguishable:

- save some additional metadata along the weights, such as the class name or parameters
  - we cannot expect that from existing weight files
  - if we want consistency, we'd have to do it for all layers - quite a big change
- change the shape saved from `GRU(reset_after=True)` so that it's a bit different from `CuDNNGRU` (an ugly hack)
- try to infer the from the weight contents - IMHO impossible

Reply:
It seems like the `weights` argument for `preprocess_weights_for_loading` is not really a list of numpy arrays. It's a list of `HDF5 dataset`. When I call `print(weights)` on my machine, I saw something like:
```
[<HDF5 dataset "kernel:0": shape (100, 96), type "<f4">,
 <HDF5 dataset "recurrent_kernel:0": shape (32, 96), type "<f4">,
 <HDF5 dataset "bias:0": shape (192,), type "<f4">]
```

Hence there would be some metadata attached to the weights. Specifically, with `print([x.name for x in weights])`, the output is:
```
['/model_weights/cu_dnngru_1/cu_dnngru_1/kernel:0',
 '/model_weights/cu_dnngru_1/cu_dnngru_1/recurrent_kernel:0',
 '/model_weights/cu_dnngru_1/cu_dnngru_1/bias:0']
```

However, I'm really not sure if this is a behavior that we could rely on (i.e., is it a consistent behavior for different versions of h5py, different versions of Keras, python, OS, ...). Also, it might not pass some existing tests since the tests are written under the assumption that the input is a list of arrays.

Reply:
What's also missing are two other cases:

- `GRU(reset_after=True)` -> `CuDNNGRU` - needs reverse conversion
- `GRU(reset_after=True)`/`CuDNNGRU` <-> `GRU(reset_after=False)` - incompatible, show fail fast with a meaningful message

It means that we could also possibly support conversion from LSTM to CuDNNLSTM (in a separate PR).

Reply:
I agree. This solution sounds good to me.